Contact form plugin

yucheng11122017 opened the pull request:
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [X] Feature addition or enhancement
- [ ] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

Closes #2095

<!--
  If this pull request is addressing an issue, link to the issue: #xxx

  If this pull request completely addresses an issue, use one of the closing keywords: "Fixes #xxx" or "Resolves #xxx"
  https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

  Otherwise, elaborate further on the rationale of this pull request as needed
-->

**Overview of changes:**
Add a web3Form plugin to allow users to create forms which uses the web3Form api 

Supports 
- Default Contact Us form 
- Standard inputs (name, email, message, submit button)

**Anything you'd like to highlight/discuss:**
In terms of number of submissions per month, web 3 form free plan seems to be the best so far with 250 submissions per month. However it only allows one email target vs [Formbold ](https://formbold.com/pricing) with 2 email target but 100 submission / month.
For now, I think web3form offers the best free plan. 

**Testing instructions:**

<!--
  Any special testing instructions **not including** our automated tests.
-->

**Proposed commit message: (wrap lines at 72 characters)**
```Add web 3 form plugin```

<!--
  See this link for more info on how to write a good commit message:
  https://se-education.org/guides/conventions/git.html

  As best as possible, write a succinct commit title in 50 characters

|---------This is the width of 50 chars----------|
|-----------This is the width of 72 chars for your reference-----------|
-->

---

**Checklist:** :ballot_box_with_check:

<!-- Leave non-applicable items unchecked -->

- [X] Updated the documentation for feature additions and enhancements
- [X] Added tests for bug fixes or features
- [X] Linked all related issues
- [X] No unrelated changes <!-- It's tempting, but increases the reviewer's work, and really pollutes the commit history =( -->

<!--
  We'll try our best to get to your PR within a week.
  If we haven't gotten to it then, or if your pull request resolves an urgent item, feel free to give us a ping!
-->

jovyntls commented on packages/core/src/plugins/web3Form.ts:
```diff
@@ -0,0 +1,150 @@
+    $node.attr('type', 'info');
+  }
+  $node.attr('no-icon', '');
+  let header = '<h2>Contact Us</h2>';
```
Hmm I'm not sure about having a (undocumented) default header here. On one hand it makes sense since this form is mostly used for "Contact Us" purposes; on the other hand, it's slightly weird to have to put `header=""` if you don't want a header. 
Either direction sounds good to me, wdyt?

yucheng11122017 replied:
Oh that's a good point. I think it would make more sense to remove the header in that case. Thank you!

yucheng11122017 replied:
I think it would be good to have it only if 1. its a default contact form and users do not specify a header. If they do, then the default header will be overriden

jovyntls reviewed (commented):
Tested and the form works smoothly for me! Nice implementation 🎉 

Mostly nits on documentation! 
The form submissions emails ended up in junk/spam for me -  not sure if adding a note on this in the UG would be suitable?

jovyntls commented on docs/userGuide/plugins/web3Form.md:
```diff
@@ -0,0 +1,89 @@
+  ],
+  "pluginsContext": {
+    "web3Form": {
+      "accesskey": "YOUR_WEB_3_FORM_ACCESS_KEY", // replace with your Web3Form access key 
```
Also is this code block language is labelled `js` instead of `json` to support comments? 🤔

yucheng11122017 replied:
Opps sorry I copied it from documentation on other plugins which was using js.

yucheng11122017 replied:
I believe so. Changing it to JSON will get warnings regarding comments and regarding the trailing commas in the start

jovyntls replied:
Ah I see, in that case I think it's fine to keep it as `js` to maintain consistency with the other plugins especially since the difference in syntax highlighting would be visible to users

damithc commented on docs/userGuide/plugins/web3Form.md:
```diff
@@ -0,0 +1,89 @@
+
+This plugin allows you to create forms whose response will be sent directly to your email, using the [Web3Form](https://web3forms.com/) API. 
+
+To set it up, get an access key from [Web3Forms](https://web3forms.com/). Then add`web3Form` to your site's plugin, and add the `accessKey` parameter via the `pluginsContext`.
```
@yucheng11122017 I assume this access key is not a 'secret' key?

yucheng11122017 replied:
Urm not exactly? It's linked to the user's email. web3forms uses it to redirect the form submissions to the user's email.

jovyntls replied:
Is there potential for misuse of the key? My current understanding is that if someone wants to abuse this key, they could use it to direct the submissions from a different form to the user's email address, but this would have the same effect as simply spamming junk submissions to the original form so it doesn't give a malicious user any additional tool for abuse. 
Unless the key can be used to reveal the user's actual email (which the user might have intended to keep private)?

damithc replied:
Should be fine, if the key is not meant to be a 'secret' key that should not be made public.

yucheng11122017 commented:
Thank you for reviewing the PR @jovyntls :)
> The form submissions emails ended up in junk/spam for me - not sure if adding a note in the UG on checking spam folders would be suitable?
I have added a warning the documentation. Also added a warning that they are limited to 250 submissions per month if they are using the free plan

tlylt commented on docs/userGuide/plugins/web3Form.md:
```diff
@@ -0,0 +1,94 @@
+
+This plugin allows you to create forms whose response will be sent directly to your email, using the [Web3Form](https://web3forms.com/) API. 
+<box type="warning" seamless>
+Emails from Web3Forms often end up in spam / junk mail.
```
```suggestion
Emails from Web3Forms may end up in spam/junk mail.
```

tlylt commented on docs/userGuide/plugins/web3Form.md:
```diff
@@ -0,0 +1,94 @@
+<include src="codeAndOutput.md" boilerplate>
+<variable name="highlightStyle">html</variable>
+<variable name="code">
+<web-3-form default>
+</web-3-form>
```
```suggestion
<web-3-form default></web-3-form>
```

Does this work if we write `<web-3-form default />` ? If so perhaps we can simplify this and other similar cases in the docs.

tlylt commented on docs/userGuide/plugins/web3Form.md:
```diff
@@ -0,0 +1,94 @@
+
+**Web3Form plugin supports customised header and styles**
+
+Options from the [box component](https://markbind.org/userGuide/components/presentation.html#boxes) are supported here.
```
use a relative reference to avoid linking out of the currently deployed site, i.e from deploy preview to markbind.org, or from markbind-master to markbind.org

tlylt commented on docs/userGuide/plugins/web3Form.md:
```diff
@@ -0,0 +1,94 @@
+The free plan from Web3Form only allows 250 submissions per month.
+</box>
+
+To set it up, get an access key from [Web3Forms](https://web3forms.com/). Then add `web3Form` to your site's plugin, and add the `accessKey` parameter via the `pluginsContext`.
```
We should mention whether the access key is ok to be exposed or at least make known that it is exposed, given that it's a working form without a backend. I tried looking at the web3forms docs and it doesn't seem to mention it at all. Perhaps it's due to the fact that the project seems really new...... Nonetheless, we should do a bit of research and put a note about this in our docs, explaining:
- whether this key is exposed to public
- whether such exposure is risky and has any consequences

yucheng11122017 replied:
The key will be exposed to the public since users can view the API call to web3forms and see the access key 
ie Developer tools -> network -> payload 
![image](https://user-images.githubusercontent.com/77204346/223029160-941a947e-8c2c-422b-ab6c-07e24f1653f2.png)
It is [mentioned](https://docs.web3forms.com/getting-started/installation#:~:text=%3C!%2D%2D%20REQUIRED%3A%20Your%20Access%20key%20here.%20Don%27t%20worry%20this%20can%20be%20public%20%2D%2D%3E) that its ok for the access key to be public. I emailed them to ask why and they said 

> Yes, it can be public.
> 
> The access key is an alias to your email. This is same as revealing email in public but one step harder. 
> 
> The people who get the access key can only used to email you. Nothing else. We also have spam protection + reCaptcha & domain protection in the Pro plan.

The problem with exposing the key is that other forms can use it and attackers might use it to DOS especially since only 250 submission are allowed per month as mentioned by @jovyntls. 

I will add the above to the documentation!

tlylt replied:
Is this in response to the question above:
![image](https://user-images.githubusercontent.com/41845017/226260242-0d82b181-a3c5-48b3-8538-0170628fc3ce.png)

?

Can we re-word it such that it says something along the line of
- generally ok to be public as shared by the team behind it (the reason you gave here)
- however, some potential problem is xxxx

This is because, at the moment, I'm not sure what to make of the warning after reading it...-> it seems to suggest that I shouldn't expose the access key given all the negatives

yucheng11122017 replied:
Thanks for the comments @tlylt, updated in [eef1b38](https://github.com/MarkBind/markbind/pull/2191/commits/eef1b38665d42a54440dfd7bb1d401def31fef98)

tlylt reviewed (changes requested):
Thank you @yucheng11122017 for the PR. Left some comments. Also, I think the suggestions from Jovyn are not yet resolved (FYI, in case you have overlooked them).

yucheng11122017 commented on packages/core/src/plugins/web3Form.ts:
```diff
@@ -0,0 +1,150 @@
+  const nodeAttribs = node.attribs;
+  if (!nodeAttribs) return;
+  delete nodeAttribs.default;
+  delete nodeAttribs.header;
```
Hi @jovyntls, is the current implementation ok? 
I can't remove the guard clause without triggering warnings

```
function deleteWeb3FormAttributes(node: cheerio.Element & DomElement) {
  if (!node.attribs) return;
  delete node.attribs.default;
  delete node.attribs.header;
}
```

jovyntls commented on docs/userGuide/plugins/web3Form.md:
```diff
@@ -0,0 +1,108 @@
+</box>
+
+<box type="warning" seamless>
+The free plan from Web3Form only allows 250 submissions per month.
```
```suggestion
The free plan from Web3Forms only allows 250 submissions per month.
```
Nit, can we standardise to use "Web3Forms" throughout? Same for the rest of the docs

jovyntls replied:
```suggestion
The free plan from Web3Form allows up to 250 submissions per month.
```
Another nit, to sound more optimistic 😄

jovyntls commented on docs/userGuide/plugins/web3Form.md:
```diff
@@ -0,0 +1,108 @@
+
+This plugin allows you to create forms whose response will be sent directly to your email, using the [Web3Form](https://web3forms.com/) API. 
+<box type="warning" seamless>
+Emails from Web3Forms may end up in spam/junk mail.
```
Thanks for adding this in! Could we rephrase this as an actionable tip for users? e.g. "Don't forget to check your spam/junk folder if your form responses aren't appearing in your inbox!" 
We could also use a `tip` type for the `<box>` to reduce the number of consecutive warning messages :)

jovyntls commented on packages/core/src/plugins/web3FormAssets/web-3-form.css:
```diff
@@ -0,0 +1,18 @@
+.form-container {
+    border-radius: 5px;
+    background-color: #00b0f0;
```
I'm guessing this background colour is meant to mirror the `primary` type box's background colour (correct me if I'm wrong!). Is it possible to apply this using the `box` container instead of hardcoding it in the CSS? This would have the benefit of being explicit about the link between the `primary` theme and this form's default colour, and also avoid future issues if there are changes in the Bootstrap themes (e.g. Bootstrap migrations).

yucheng11122017 replied:
Sorry removed! I did this before moving over to markbind's box :sweat_smile:

jovyntls commented on packages/core/src/plugins/web3Form.ts:
```diff
@@ -0,0 +1,153 @@
+    return;
+  }
+  const $node = cheerio(child);
+  switch (child.name.toLowerCase()) {
```
Valid edge case, but I believe conversion to lowercase is not necessary as it's already done in [NodeProcessor](https://github.com/yucheng11122017/markbind/blob/f12518e6281aceae8fca50d9e336cb17df1cdf90/packages/core/src/html/NodeProcessor.ts#L321) before the pluginManager calls `processNode` :)

jovyntls reviewed (changes requested):
Thanks for making all the previous changes @yucheng11122017 🔥  
Added some more comments! And nits on documentation

jovyntls commented on packages/core/src/plugins/web3FormAssets/web-3-form.css:
```diff
@@ -0,0 +1,12 @@
+input,
+select,
+textarea {
```
One last nit, do the Bootstrap form classes work well here or would it be more cumbersome than using CSS? https://getbootstrap.com/docs/5.1/forms/overview/#overview

yucheng11122017 replied:
Mmm they definitely work but if bootstrap is used the problem is for custom inputs (ie not using `name-input` or `message-input` etc) they would not be styled the same way since users may not have added the bootstrap class. 

Eg. something like this 
```
<web-3-form>
  <label for="fname">First name:</label><br>
  <input type="text" id="fname" name="fname"><br>
  <label for="lname">Last name:</label><br>
  <input type="text" id="lname" name="lname">
</web-3-form>
```

I can change it to add the bootstrap class for the custom inputs if that is better?

jovyntls replied:
Sorry I missed this! I see, in that case the Bootstrap classes are quite cumbersome - it seems that CSS is a better solution 👍

tlylt commented:
![image](https://user-images.githubusercontent.com/41845017/226261330-783ab02d-cdb3-4ad0-bff9-1aa474051824.png)
The page nav is corrupted by the example

yucheng11122017 commented:
> ![image](https://user-images.githubusercontent.com/41845017/226261330-783ab02d-cdb3-4ad0-bff9-1aa474051824.png) The page nav is corrupted by the example

@tlylt I changed some of the headings to bold instead in [7a149d9](https://github.com/MarkBind/markbind/pull/2191/commits/7a149d92e3b963b8b3c7a75056c4fc534ef6a0f9). But for the default contact form the heading is `<h2>` which will be picked up by the page nav. Is that ok?

tlylt commented:
> > ![image](https://user-images.githubusercontent.com/41845017/226261330-783ab02d-cdb3-4ad0-bff9-1aa474051824.png) The page nav is corrupted by the example
> 
> @tlylt I changed some of the headings to bold instead in [7a149d9](https://github.com/MarkBind/markbind/pull/2191/commits/7a149d92e3b963b8b3c7a75056c4fc534ef6a0f9). But for the default contact form the heading is `<h2>` which will be picked up by the page nav. Is that ok?

Hi @yucheng11122017, I just discussed with @jovyntls, I think if we can change the “contact us" default from h2 to a div with the relevant bootstrap classes to style it, it would prevent the issue from occurring. What do you think about this?

jovyntls reviewed (dismissed):
LGTM! Thanks again @yucheng11122017 

> Yeah ok sounds good! There is nothing that disables the page-nav from rendering some headings right haha

Nope as far as I know! Might be a possible feature request but I don't think there's much utility to doing so (especially since using Bootstrap styles would be a possible quick workaround for authors who wish to do so)

(Edited the PR description to close the original issue with keywords!)

tlylt commented on packages/core/src/plugins/web3FormAssets/web-3-form.css:
```diff
@@ -0,0 +1,12 @@
+input,
```
Hi @yucheng11122017, these CSS selectors are too generic, it causes any input etc elements to have unwanted effects. The most obvious one that caught my attention is the increased in margin for our header. See the preview:
![image](https://user-images.githubusercontent.com/41845017/227410354-a6f4c238-c69e-4f5f-b9bd-17d68343dc5e.png)

Current working version is:
![image](https://user-images.githubusercontent.com/41845017/227410433-e82932f6-740f-407a-a6bf-c6e74296df1a.png)

yucheng11122017 commented:
Hi @tlylt, thanks for bringing that up! I changed the selector to only affect the inputs in the web3Form.

jovyntls commented on packages/core/src/plugins/web3Form.ts:
One last nit, can we use the newly introduced node types (from #2201 - MbNode, NodeOrText etc) wherever possible instead of `cheerio.Element` and `DomElement`?

yucheng11122017 replied:
Added in [ac99277](https://github.com/MarkBind/markbind/pull/2191/commits/ac99277ae0285f48252113d63b809fda4079eead)! :)

tlylt replied:
Hi @yucheng11122017, just want to share that besides changing the node type, there might be a need to adjust the code accordingly. So in this case, some of the lines do not make sense anymore:
e.g. these few lines are not needed anymore because `attribs` will always exist
![image](https://user-images.githubusercontent.com/41845017/227460207-1c41752c-bbb8-451f-9333-aee4f8e88cd8.png)

In the future do take note of the general lesson here as well: the review comment may be specific and points out one thing, that doesn't mean other relevant code changes will not be needed. So relooking at what other potential changes are necessary would help a lot.

jovyntls reviewed (commented):
Didn't catch this earlier (my bad)!

jovyntls reviewed (approved):
LGTM!! 🥳🥳🥳

tlylt reviewed (approved):
@yucheng11122017 Try merging this (note that this is not typescript PR, so it should be a squash commit strategy)
